storage/api/src/main/java/feast/storage/api/retrieval/OnlineRetriever.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2020 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.storage.api.retrieval;
+
+import feast.serving.ServingAPIProto.GetOnlineFeaturesRequest;
+import feast.types.FeatureRowProto;
+import java.io.Serializable;
+import java.util.List;
+
+/** Interface for implementing user defined retrieval functionality from Online stores. */
+public interface OnlineRetriever<T> extends Serializable {
+
+  /**
+   * Get all values corresponding to the request.
+   *
+   * @param request Online features request containing the list of features and list of entity rows
+   * @return list of {@link T}
+   */
+  List<T> getOnlineFeatures(GetOnlineFeaturesRequest request);
+
+  /**
+   * Checks whether the response is empty, i.e. feature does not exist in the store
+   *
+   * @param response {@link T}
+   * @return boolean
+   */
+  boolean isEmpty(T response);
+
+  /**
+   * Parse response from data store to FeatureRow
+   *
+   * @param response {@link T}
+   * @return {@link FeatureRowProto.FeatureRow}
+   */
+  FeatureRowProto.FeatureRow parseResponse(T response);
```

Comment:
> It is less generic than the BatchRetriever interface so that implementations don't have to implement the code tracking missing/stale features.

Would you mind fleshing out a typical flow here? It's not clear to me what the different `T` are that we would want to return from `getOnlineFeatures`

Reply:
T is store specific response to a lookup. Refer to @smadarasmi's implementation [here](https://github.com/gojek/feast/blob/v0.3-branch/serving/src/main/java/feast/serving/service/OnlineServingService.java#L176) for the example flow

Reply:
Had a look at it now. I am not 100% sure what value this brings. It seems like what we are trying to do is externalize the looping to the layer above the store in order to make it reusable?

I think this makes the code harder to read and the interface more complex, for very little gain. At the end of the day the implementer still needs to write `isEmpty()` and `parseResponse()` themselves, but are these methods ever used outside of this specific `getFeatures` request?

The outer layer cannot do anything with `T` except treat it as a list and send it back to the store until the store finally gives it what it wants (`FeatureRows`).

Or am I missing something? cc @ches @smadarasmi

Reply:
If what we are looking for is metadata about missing or stale features then I would argue it is better to have a different (but specific) response type from getOnlineFeatures so that we get that information from the store in a way that the layer about the store can understand.

Reply:
made changes, can you have a look? I'm still partial to outsourcing most of the work to the external loop since it's significant boilerplate (particularly the logging of metrics), but I've moved the parsing/isEmpty logic to a specific object.

Reply:
Cool. This would be easy to refactor later anyways.